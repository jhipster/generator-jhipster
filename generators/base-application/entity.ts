import { kebabCase, lowerFirst, snakeCase, startCase, upperFirst } from 'lodash-es';
import pluralize from 'pluralize';

import type { DerivedPropertiesOnlyOf } from '../../lib/command/types.ts';
import type { FieldType } from '../../lib/jhipster/field-types.ts';
import { BlobTypes, fieldTypesValues } from '../../lib/jhipster/field-types.ts';
import { type ValidationType, validationTypes } from '../../lib/jhipster/index.ts';
import type { Entity as BaseEntity } from '../../lib/jhipster/types/entity.ts';
import type { Field as BaseField } from '../../lib/jhipster/types/field.ts';
import type { Relationship as BaseRelationship } from '../../lib/jhipster/types/relationship.ts';
import { buildMutateDataForProperty } from '../../lib/utils/derived-property.ts';
import type { MutateDataParam, MutateDataPropertiesWithRequiredProperties } from '../../lib/utils/object.ts';

import { isFieldEnumType } from './internal/types/field-types.ts';
import type { FakerWithRandexp } from './support/faker.ts';

type Property = {
  propertyName: string;

  propertyNameCapitalized?: string;
  propertyNameUpperSnakeCase?: string;
  propertyApiDescription?: string;

  skipClient?: boolean;
  skipServer?: boolean;
};

const mutateProperty = {
  propertyNameCapitalized: ({ propertyName }) => upperFirst(propertyName),
  propertyNameUpperSnakeCase: ({ propertyName }) => snakeCase(propertyName).toUpperCase(),
} as const satisfies MutateDataParam<Property>;

type BaseApplicationAddedFieldProperties = DerivedPropertiesOnlyOf<'fieldType', FieldType> &
  DerivedPropertiesOnlyOf<'fieldValidation', ValidationType> & {
    path?: string[];

    fieldNameCapitalized: string;
    fieldNameHumanized: string;
    fieldNameUnderscored: string;
    fieldTranslationKey?: string;

    fieldApiDescription?: string;

    enumFileName?: string;
    enumValues?: { name: string; value: string }[];

    // Validation
    fieldValidate?: boolean;
    unique?: boolean;
    maxlength?: number;

    /**
     * Faker template passed to `faker.helpers.fake()`.
     * @see https://fakerjs.dev/api/helpers.html#fake
     */
    fakerTemplate?: string;

    // Temporary fields for Faker
    uniqueValue?: any[];

    generateFakeDataFromPattern?: () => string | undefined;

    generateFakeData?: (type?: 'csv' | 'cypress' | 'json-serializable' | 'ts') => any;

    transient?: boolean;
    id?: boolean;
    autoGenerate?: boolean;
    readonly?: boolean;
    sequenceGeneratorName?: string;

    // Blob
    fieldWithContentType?: boolean;
    contentTypeFieldName?: string;

    fieldTypeTimed?: boolean;

    blobContentTypeText?: boolean;
    blobContentTypeImage?: boolean;
    blobContentTypeAny?: boolean;

    fieldTypeBytes?: boolean;
    // Derived properties
    fieldTypeBinary?: boolean;

    /** @deprecated */
    fieldValidationMinLength: boolean;
    /** @deprecated */
    fieldValidationMaxLength: boolean;
    /** @deprecated */
    fieldValidationMinBytes: boolean;
    /** @deprecated */
    fieldValidationMaxBytes: boolean;

    relatedByOtherEntity?: boolean;

    enumInstance?: string;
    builtIn?: boolean;
  };

export type Field = Property &
  Omit<BaseField, 'fieldType'> &
  BaseApplicationAddedFieldProperties &
  (
    | {
        fieldType: string;
        fieldIsEnum: true;
      }
    | {
        fieldType: FieldType;
        fieldIsEnum: false;
      }
  );

export const mutateField = {
  __override__: false,
  fieldIsEnum: data => isFieldEnumType(data),
  ...buildMutateDataForProperty('fieldType', Object.values(fieldTypesValues), { anyData: true }),
  fieldTypeBytes: ({ fieldTypeByte }) => fieldTypeByte,
  ...buildMutateDataForProperty('fieldTypeBlobContent', Object.values(BlobTypes), { prefix: 'blobContentType' }),

  path: ({ fieldName }) => [fieldName],
  propertyName: ({ fieldName }) => fieldName,
  ...mutateProperty,

  fieldNameCapitalized: ({ fieldName }) => upperFirst(fieldName),
  fieldNameUnderscored: ({ fieldName }) => snakeCase(fieldName),
  fieldNameHumanized: ({ fieldName }) => startCase(fieldName),

  ...buildMutateDataForProperty('fieldValidateRules', Object.values(validationTypes), {
    prefix: 'fieldValidation',
    array: true,
    valCheck: (data, value) => data.fieldValidateRules?.includes(value) ?? false,
  }),
  fieldValidationMaxLength: ({ fieldValidationMaxlength }) => fieldValidationMaxlength,
  fieldValidationMinLength: ({ fieldValidationMinlength }) => fieldValidationMinlength,
  fieldValidationMinBytes: ({ fieldValidationMinbytes }) => fieldValidationMinbytes,
  fieldValidationMaxBytes: ({ fieldValidationMaxbytes }) => fieldValidationMaxbytes,
} as const satisfies MutateDataPropertiesWithRequiredProperties<MutateDataParam<Field>, BaseApplicationAddedFieldProperties>;

export type DerivedField<E extends Entity = Entity, F extends Field = Entity['fields'][number]> = F & {
  derived: true;
  originalField: F;
  derivedEntity: E;
};

type BaseApplicationAddedRelationshipProperties = DerivedPropertiesOnlyOf<
  'relationship',
  'LeftSide' | 'RightSide' | 'ManyToOne' | 'OneToMany' | 'OneToOne' | 'ManyToMany'
> &
  Property & {
    relationshipNameCapitalized: string;
    collection: boolean;

    /**
     * A persistable relationship means that the relationship will be updated in the database.
     */
    persistableRelationship?: boolean;

    id?: boolean;
    /** @deprecated */
    ownerSide?: boolean;
    relationshipEagerLoad?: boolean;
    relationshipRequired?: boolean;

    relationshipFieldName: string;
    relationshipFieldNamePlural: string;
    relationshipNamePlural: string;
    relationshipNameHumanized: string;

    relationshipIgnoreBackReference?: boolean;

    relationshipValidate?: boolean;
    relationshipValidateRules?: string[];
  };

/**
 * Represents a relationship with an otherRelationship.
 * Interface is used to allow `this` type in the otherRelationship.
 */
export interface Relationship extends BaseApplicationAddedRelationshipProperties, BaseRelationship {
  otherRelationship: this;
}

export const mutateRelationship = {
  __override__: false,
  ...buildMutateDataForProperty('relationshipSide', ['left', 'right'], { prefix: 'relationship', suffix: 'Side' }),
  ...buildMutateDataForProperty('relationshipType', ['one-to-one', 'one-to-many', 'many-to-one', 'many-to-many'], {
    prefix: 'relationship',
  }),
  collection: ({ relationshipType }) => relationshipType === 'one-to-many' || relationshipType === 'many-to-many',

  relationshipFieldName: ({ relationshipName }) => lowerFirst(relationshipName),
  relationshipFieldNamePlural: ({ relationshipFieldName }) => pluralize(relationshipFieldName!),
  relationshipNamePlural: ({ relationshipName }) => pluralize(relationshipName),
  relationshipNameCapitalized: ({ relationshipName }) => upperFirst(relationshipName),
  relationshipNameHumanized: ({ relationshipName }) => startCase(relationshipName),

  propertyName: ({ collection, relationshipFieldName, relationshipFieldNamePlural }) =>
    collection ? relationshipFieldNamePlural! : relationshipFieldName!,

  ...mutateProperty,
} as const satisfies MutateDataPropertiesWithRequiredProperties<MutateDataParam<Relationship>, BaseApplicationAddedRelationshipProperties>;

/**
 * Represents a relationship with an otherEntity, where the relationship is extended with the other entity.
 * Utility type to workaround https://github.com/Microsoft/TypeScript/issues/24364.
 */
export type RelationshipWithEntity<R, E extends BaseEntity> = R & {
  otherEntity: E;
  relatedField: Exclude<E['fields'], undefined>[number];
  otherEntityUser: boolean;
};

export const mutateRelationshipWithEntity = {
  __override__: false,
  otherEntityField: data => data.otherEntity?.primaryKey?.name,
  // let ownerSide true when type is 'many-to-one' for convenience.
  // means that this side should control the reference.
  ownerSide: data =>
    Boolean(data.otherEntity.embedded || data.relationshipManyToOne || (data.relationshipLeftSide && !data.relationshipOneToMany)),
  persistableRelationship: ({ ownerSide }) => ownerSide!,
  otherEntityUser: ({ otherEntityName }) => otherEntityName.toLowerCase() === 'user',
} satisfies MutateDataParam<RelationshipWithEntity<Relationship, Entity>>;

export type PrimaryKey<F extends Field = Field> = {
  name: string;
  nameCapitalized: string;
  hibernateSnakeCaseName: string;
  fields: F[];
  relationships: any[];

  type: FieldType;
  composite: boolean;
  derived: boolean;
  javaValueGenerator?: string;
  javaBuildSpecification?: string;

  tsSampleValues?: (string | number)[];
  javaSampleValues?: string[];

  hasUUID?: boolean;
  hasLong?: boolean;
  hasInteger?: boolean;
  typeUUID?: boolean;
  typeString?: boolean;
  typeLong?: boolean;
  typeInteger?: boolean;
  /** @deprecated replace with technology-specific implementation */
  typeNumeric?: boolean;

  derivedFields?: (F & {
    originalField: F;
    derived: boolean;
  })[];
  ownFields?: F[];

  tsType?: string;
  autoGenerate?: boolean;
  /** @deprecated */
  ids: any[];
};

type BaseApplicationAddedEntityProperties = {
  entityAuthority?: string;
  entityReadAuthority?: string;

  /** @experimental */
  auditableEntity?: boolean;
  builtIn?: boolean;
  builtInUser?: boolean;
  builtInUserManagement?: boolean;
  builtInAuthority?: boolean;
  adminEntity?: boolean;
  hasCyclicRequiredRelationship?: boolean;

  entityNameCapitalized: string;
  entityNamePlural: string;
  entityNamePluralizedAndSpinalCased: string;
  entityInstancePlural: string;
  entityInstance: string;

  // TODO rename to entityNameHumanized
  entityClassHumanized: string;
  // TODO rename to entityNamePluralHumanized
  entityClassPluralHumanized: string;

  resetFakerSeed?(suffix?: string): void;
  generateFakeData?: (type?: any) => any;
  faker?: FakerWithRandexp;

  anyFieldIsBigDecimal?: boolean;
  /**
   * Any file is of type Bytes or ByteBuffer
   */
  anyFieldIsBlobDerived?: boolean;
  /**
   * Any field is of type ZonedDateTime, Instant or LocalDate
   */
  anyFieldIsDateDerived?: boolean;
  anyFieldIsDuration?: boolean;
  anyFieldIsInstant?: boolean;
  anyFieldIsLocalDate?: boolean;
  anyFieldIsLocalTime?: boolean;
  /**
   * Any field is of type ZonedDateTime or Instant
   */
  anyFieldIsTimeDerived?: boolean;
  anyFieldIsUUID?: boolean;
  anyFieldIsZonedDateTime?: boolean;

  anyFieldHasDocumentation?: boolean;
  anyFieldHasImageContentType?: boolean;
  anyFieldHasTextContentType?: boolean;
  /**
   * Any field has image or any contentType
   */
  anyFieldHasFileBasedContentType?: boolean;

  /**
   * Any relationship is required or id
   */
  anyRelationshipIsRequired?: boolean;
  hasRelationshipWithBuiltInUser?: boolean;

  /** Properties from application required for entities published through gateways */
  useMicroserviceJson?: boolean;
  microserviceAppName?: string;
  applicationType?: string;
  microfrontend?: boolean;
  skipUiGrouping?: boolean;
};

/**
 * Represents an entity with its relationships, where the relationships are extended with the other entity.
 * Interface is used to allow `this` type in the relationships.
 */
export interface Entity<F extends Field = Field, R extends Relationship = Relationship>
  extends BaseApplicationAddedEntityProperties,
    Omit<Required<BaseEntity<F>>, 'relationships'> {
  relationships: RelationshipWithEntity<R, this>[];
  otherRelationships: R[];

  primaryKey?: PrimaryKey<F>;

  changelogDateForRecent: any;
}

export const mutateEntity = {
  __override__: false,
  entityNameCapitalized: ({ name }) => upperFirst(name),
  entityNamePlural: ({ name }) => pluralize(name),
  entityNamePluralizedAndSpinalCased: ({ entityNamePlural }) => kebabCase(entityNamePlural),
  entityInstance: ({ name }) => lowerFirst(name),
  entityInstancePlural: ({ entityNamePlural }) => lowerFirst(entityNamePlural),
  entityAuthority: ({ adminEntity }) => (adminEntity ? 'ROLE_ADMIN' : undefined),

  entityClassHumanized: ({ entityNameCapitalized }) => startCase(entityNameCapitalized),
  entityClassPluralHumanized: ({ entityNamePlural }) => startCase(entityNamePlural),
} as const satisfies MutateDataPropertiesWithRequiredProperties<MutateDataParam<Entity>, BaseApplicationAddedEntityProperties>;
