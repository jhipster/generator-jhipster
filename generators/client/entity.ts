/**
 * Copyright 2013-2025 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { kebabCase, upperFirst } from 'lodash-es';
import pluralize from 'pluralize';

import type { MutateDataParam, MutateDataPropertiesWithRequiredProperties } from '../../lib/utils/object.ts';
import { normalizePathEnd } from '../../lib/utils/path.ts';
import { upperFirstCamelCase } from '../../lib/utils/string.ts';
import type { Relationship as BaseApplicationRelationship } from '../base-application/types.d.ts';
import type { Entity as CommonEntity, Field as CommonField, Relationship as CommonRelationship } from '../common/types.ts';
import type { Entity as LanguagesEntity, Field as LanguagesField, Relationship as LanguagesRelationship } from '../languages/types.d.ts';

import { getTypescriptType } from './support/types-utils.ts';

// DerivedPropertiesOnlyOf<'fieldTsType', FieldTsType> &
type ClientAddedFieldProperties = {
  tsType: string;
};

export const mutateField = {
  __override__: false,
  tsType: ({ fieldType, fieldIsEnum }) => (fieldIsEnum ? fieldType : getTypescriptType(fieldType)),

  // ...buildMutateDataForProperty('tsType', fieldTsTypes, { prefix: 'fieldTsType' }),
} as const satisfies MutateDataPropertiesWithRequiredProperties<MutateDataParam<Field>, ClientAddedFieldProperties>;

export type Field = CommonField & LanguagesField & ClientAddedFieldProperties;

export interface Relationship extends CommonRelationship, LanguagesRelationship {}

type ClientAddedEntityProperties = {
  entityFileName: string;
  entityFolderName: string;
  entityModelFileName: string;
  entityPluralFileName: string;
  entityServiceFileName: string;

  /** Generate only the model at client side for relationships. */
  entityClientModelOnly?: boolean;
  entityTsName: string;
  entityAngularName: string;
  entityAngularNamePlural: string;
  entityReactName: string;
  entityStateName: string;
  entityUrl: string;
  entityPage?: string;

  tsKeyType?: string;
  tsSampleWithPartialData?: string;
  tsSampleWithRequiredData?: string;
  tsSampleWithFullData?: string;
  tsSampleWithNewData?: string;
  tsPrimaryKeySamples?: string[];

  entityAngularJSSuffix: string;
};

export const mutateEntity = {
  __override__: false,
  entityAngularJSSuffix: data => {
    const entityAngularJSSuffix = data.entityAngularJSSuffix ?? data.angularJSSuffix ?? '';
    return entityAngularJSSuffix.startsWith('-') || !entityAngularJSSuffix ? entityAngularJSSuffix : `-${entityAngularJSSuffix}`;
  },
  entityTsName: data => upperFirst(data.entityNameCapitalized) + upperFirstCamelCase(data.entityAngularJSSuffix!),
  entityFileName: data => kebabCase(data.entityNameCapitalized + upperFirst(data.entityAngularJSSuffix)),
  entityFolderName: data => `${normalizePathEnd(data.clientRootFolder)}${data.entityFileName}`,
  entityModelFileName: data => data.entityFolderName,
  entityPluralFileName: data => `${data.entityNamePluralizedAndSpinalCased}${data.entityAngularJSSuffix}`,
  entityServiceFileName: data => data.entityFileName,
  entityStateName: data => kebabCase(data.entityTsName),
  entityUrl: data => data.entityStateName,

  entityAngularName: data => data.entityTsName,
  entityAngularNamePlural: data => pluralize(data.entityAngularName),
  entityReactName: data => data.entityTsName,
} as const satisfies MutateDataPropertiesWithRequiredProperties<MutateDataParam<Entity>, ClientAddedEntityProperties>;

export interface Entity<F extends Field = Field, R extends BaseApplicationRelationship = BaseApplicationRelationship>
  extends CommonEntity<F, R>,
    LanguagesEntity<F, R>,
    ClientAddedEntityProperties {}
