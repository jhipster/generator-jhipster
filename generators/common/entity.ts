/**
 * Copyright 2013-2025 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { CommonDBTypes, RelationalOnlyDBTypes } from '../../lib/jhipster/field-types.ts';
import { buildMutateDataForProperty } from '../../lib/utils/derived-property.ts';
import type { MutateDataParam, MutateDataPropertiesWithRequiredProperties } from '../../lib/utils/object.ts';
import type {
  Entity as BaseApplicationEntity,
  Field as BaseApplicationField,
  Relationship as BaseApplicationRelationship,
} from '../base-application/types.ts';
import type { Entity as LanguagesEntity } from '../languages/types.ts';

const { BIG_DECIMAL, DOUBLE, FLOAT, INSTANT, INTEGER, LOCAL_DATE, LONG, STRING, UUID, ZONED_DATE_TIME, TEXT_BLOB } = CommonDBTypes;
const { BYTES, BYTE_BUFFER } = RelationalOnlyDBTypes;

type CommonAddedFieldProperties = {
  /** @deprecated replace with technology-specific implementation */
  fieldTypeTemporal: boolean;
  /** @deprecated replace with technology-specific implementation */
  fieldTypeCharSequence: boolean;
  /** @deprecated replace with technology-specific implementation */
  fieldTypeNumeric: boolean;
  /** @deprecated replace with technology-specific implementation */
  fieldSupportsSortBy?: boolean;
};

export type Field = BaseApplicationField & CommonAddedFieldProperties;

export const mutateField = {
  __override__: false,
  fieldTypeNumeric: ({ fieldType }) =>
    fieldType === INTEGER || fieldType === LONG || fieldType === FLOAT || fieldType === DOUBLE || fieldType === BIG_DECIMAL,
  fieldTypeBinary: ({ fieldType }) => fieldType === BYTES || fieldType === BYTE_BUFFER,
  fieldTypeTimed: ({ fieldType }) => fieldType === ZONED_DATE_TIME || fieldType === INSTANT,
  fieldTypeCharSequence: ({ fieldType }) => fieldType === STRING || fieldType === UUID || fieldType === TEXT_BLOB,
  fieldTypeTemporal: ({ fieldType }) => fieldType === ZONED_DATE_TIME || fieldType === INSTANT || fieldType === LOCAL_DATE,
} as const satisfies MutateDataPropertiesWithRequiredProperties<MutateDataParam<Field>, CommonAddedFieldProperties>;

export type { BaseApplicationRelationship as Relationship };

type CommonAddedEntityProperties = {
  entityApiUrl: string;
  entityApi: string;

  paginationPagination: boolean;
  paginationInfiniteScroll: boolean;
  paginationNo: boolean;
};

export const mutateEntity = {
  __override__: false,
  entityApi: ({ microserviceName }) => (microserviceName ? `services/${microserviceName.toLowerCase()}/` : ''),
  entityApiUrl: data => data.entityNamePluralizedAndSpinalCased,
  ...buildMutateDataForProperty('pagination', ['pagination', 'infinite-scroll', 'no']),
} as const satisfies MutateDataPropertiesWithRequiredProperties<MutateDataParam<Entity>, CommonAddedEntityProperties>;

export interface Entity<F extends Field = Field, R extends BaseApplicationRelationship = BaseApplicationRelationship>
  extends LanguagesEntity<F, R>,
    BaseApplicationEntity<F, R>,
    CommonAddedEntityProperties {
  restProperties?: (F | R)[];

  uniqueEnums?: F[];
}
