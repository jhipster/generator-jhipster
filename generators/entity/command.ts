/**
 * Copyright 2013-2025 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import type { JHipsterCommandDefinition } from '../../lib/command/index.js';

const command: JHipsterCommandDefinition = {
  arguments: {
    name: {
      type: String,
      required: true,
      description: 'Entity name',
    },
  },
  options: {
    regenerate: {
      description: 'Regenerate the entity without presenting an option to update it',
      type: Boolean,
      default: false,
    },
    tableName: {
      description: 'Specify table name that will be used by the entity',
      type: String,
    },
    fluentMethods: {
      description: 'Generate fluent methods in entity beans to allow chained object construction',
      type: Boolean,
    },
    angularSuffix: {
      description: 'Use a suffix to generate Angular routes and files, to avoid name clashes',
      type: String,
    },
    clientRootFolder: {
      description:
        'Use a root folder name for entities on client side. By default its empty for monoliths and name of the microservice for gateways',
      type: String,
    },
    skipUiGrouping: {
      description: 'Disables the UI grouping behaviour for entity client side code',
      type: Boolean,
    },
    skipDbChangelog: {
      description: 'Skip the generation of database changelog (liquibase for sql databases)',
      type: Boolean,
    },
    singleEntity: {
      description: 'Regenerate only a single entity, relationships can be not correctly generated',
      type: Boolean,
    },
  },
};

export default command;
