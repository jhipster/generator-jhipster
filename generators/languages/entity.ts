/**
 * Copyright 2013-2025 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { camelCase } from 'lodash-es';

import type { MutateDataParam, MutateDataPropertiesWithRequiredProperties } from '../../lib/utils/object.ts';
import type {
  Entity as BaseApplicationEntity,
  Field as BaseApplicationField,
  Relationship as BaseApplicationRelationship,
} from '../base-application/types.ts';

export type { BaseApplicationField as Field, BaseApplicationRelationship as Relationship };

type LanguagesAddedEntityProperties = {
  /** i18n variant ex: 'male', 'female' when applied */
  entityI18nVariant: string;
  entityTranslationKey: string;
  entityTranslationKeyMenu: string;
};

export interface Entity<
  F extends BaseApplicationField = BaseApplicationField,
  R extends BaseApplicationRelationship = BaseApplicationRelationship,
> extends BaseApplicationEntity<F, R>,
    LanguagesAddedEntityProperties {
  i18nKeyPrefix: string;
  i18nAlertHeaderPrefix: string;
}

export const mutateEntity = {
  __override__: false,
  entityI18nVariant: 'default',
  entityTranslationKey: data =>
    data.clientRootFolder ? camelCase(`${data.clientRootFolder}-${data.entityInstance}`) : data.entityInstance,
  entityTranslationKeyMenu: data =>
    camelCase(data.clientRootFolder ? `${data.clientRootFolder}-${data.entityNameKebabCase}` : data.entityNameKebabCase),
} as const satisfies MutateDataPropertiesWithRequiredProperties<MutateDataParam<Entity>, LanguagesAddedEntityProperties>;
