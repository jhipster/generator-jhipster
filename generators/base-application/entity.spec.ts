/**
 * Copyright 2013-2025 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { expect } from 'esmocha';

import { mutateMockedCompleteData, mutateMockedData } from '../../lib/testing/index.ts';

import { mutateEntity, mutateRelationship, mutateRelationshipWithEntity } from './entity.ts';

describe('mutation object test', () => {
  it('expects mutateRelationship to match snapshot', () => {
    expect(mutateMockedData(mutateRelationship)).toMatchSnapshot();
  });
  it("expects mutateRelationship to don't override existing properties", () => {
    expect(Object.keys(mutateMockedCompleteData(mutateRelationship))).toHaveLength(0);
  });
  it('expects mutateRelationshipWithEntity to match snapshot', () => {
    expect(mutateMockedData(mutateRelationshipWithEntity)).toMatchSnapshot();
  });
  it("expects mutateRelationshipWithEntity to don't override existing properties", () => {
    expect(Object.keys(mutateMockedCompleteData(mutateRelationshipWithEntity))).toHaveLength(0);
  });
  it('expects mutateEntity to match snapshot', () => {
    expect(mutateMockedData(mutateEntity)).toMatchSnapshot();
  });
  it("expects mutateEntity to don't override existing properties", () => {
    expect(Object.keys(mutateMockedCompleteData(mutateEntity))).toHaveLength(0);
  });
});
