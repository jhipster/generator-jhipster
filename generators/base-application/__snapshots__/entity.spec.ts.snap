// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`mutation object test expects mutateEntity to match snapshot 1`] = `
{
  "clientRootFolder": "",
  "entityAuthority": "ROLE_ADMIN",
  "entityClassHumanized": "Name",
  "entityClassPluralHumanized": "Names",
  "entityInstance": "name",
  "entityInstancePlural": "names",
  "entityNameCapitalized": "Name",
  "entityNameKebabCase": "name",
  "entityNamePlural": "names",
  "entityNamePluralizedAndSpinalCased": "names",
}
`;

exports[`mutation object test expects mutateField to match snapshot 1`] = `
{
  "blobContentTypeAny": false,
  "blobContentTypeImage": false,
  "blobContentTypeText": false,
  "fieldIsEnum": true,
  "fieldNameCapitalized": "FieldName",
  "fieldNameHumanized": "Field Name",
  "fieldNameUnderscored": "field_name",
  "fieldTypeAnyBlob": false,
  "fieldTypeBigDecimal": false,
  "fieldTypeBlob": false,
  "fieldTypeBoolean": false,
  "fieldTypeByte": false,
  "fieldTypeByteBuffer": false,
  "fieldTypeBytes": false,
  "fieldTypeDouble": false,
  "fieldTypeDuration": false,
  "fieldTypeFloat": false,
  "fieldTypeImageBlob": false,
  "fieldTypeInstant": false,
  "fieldTypeInteger": false,
  "fieldTypeLocalDate": false,
  "fieldTypeLocalTime": false,
  "fieldTypeLong": false,
  "fieldTypeString": false,
  "fieldTypeTextBlob": false,
  "fieldTypeUUID": false,
  "fieldTypeZonedDateTime": false,
  "fieldValidationMax": false,
  "fieldValidationMaxBytes": false,
  "fieldValidationMaxLength": false,
  "fieldValidationMaxbytes": false,
  "fieldValidationMaxlength": false,
  "fieldValidationMin": false,
  "fieldValidationMinBytes": false,
  "fieldValidationMinLength": false,
  "fieldValidationMinbytes": false,
  "fieldValidationMinlength": false,
  "fieldValidationPattern": false,
  "fieldValidationRequired": false,
  "fieldValidationUnique": false,
  "path": [
    "fieldName",
  ],
  "propertyName": "fieldName",
  "propertyNameCapitalized": "FieldName",
  "propertyNameUpperSnakeCase": "FIELD_NAME",
}
`;

exports[`mutation object test expects mutateRelationship to match snapshot 1`] = `
{
  "collection": false,
  "propertyName": "relationshipName",
  "propertyNameCapitalized": "RelationshipName",
  "propertyNameUpperSnakeCase": "RELATIONSHIP_NAME",
  "relationshipFieldName": "relationshipName",
  "relationshipFieldNamePlural": "relationshipNames",
  "relationshipLeftSide": false,
  "relationshipManyToMany": false,
  "relationshipManyToOne": false,
  "relationshipNameCapitalized": "RelationshipName",
  "relationshipNameHumanized": "Relationship Name",
  "relationshipNamePlural": "relationshipNames",
  "relationshipOneToMany": false,
  "relationshipOneToOne": false,
  "relationshipRightSide": false,
}
`;

exports[`mutation object test expects mutateRelationshipWithEntity to match snapshot 1`] = `
{
  "otherEntityField": undefined,
  "otherEntityUser": false,
  "ownerSide": true,
  "persistableRelationship": true,
}
`;
